File: modules/fhir-structure/src/blaze/fhir/hash.clj
(ns blaze.fhir.hash
  (:require
   [blaze.byte-buffer :as bb])
  (:import
   [blaze.fhir Hash]
   [blaze.fhir.spec.type Base]
   [com.fasterxml.jackson.core JsonGenerator]
   [com.fasterxml.jackson.databind.module SimpleModule]
   [com.fasterxml.jackson.databind.ser.std StdSerializer]
   [com.google.common.hash Hashing]
   [java.io Writer]))

(set! *warn-on-reflection* true)

(def ^:const ^long size 32)
(def ^:const ^long prefix-size Integer/BYTES)

(def deleted-hash
  "The hash of a deleted version of a resource."
  Hash/DELETED)

(defn hash? [x]
  (instance? Hash x))

(defn from-byte-buffer! [byte-buffer]
  (Hash/fromByteBuffer byte-buffer))

(defn from-hex [s]
  (Hash/fromHex s))

(defn into-byte-buffer! [byte-buffer hash]
  (.copyTo ^Hash hash byte-buffer)
  byte-buffer)

(defn prefix-into-byte-buffer!
  [byte-buffer hash]
  (bb/put-int! byte-buffer (.prefix ^Hash hash)))

(defn to-byte-array [hash]
  (-> (bb/allocate size)
      (into-byte-buffer! hash)
      (bb/array)))

(defn prefix
  "Returns the first 4 bytes of `hash`."
  [hash]
  (bit-and (.prefix ^Hash hash) 0xFFFFFFFF))

(defn prefix-from-byte-buffer!
  [byte-buffer]
  (bit-and (bb/get-int! byte-buffer) 0xFFFFFFFF))

(defn prefix-from-hex [s]
  (Long/parseLong s 16))

(def ^:private serializer
  (proxy [StdSerializer] [Hash]
    (serialize [^Hash hash ^JsonGenerator gen _]
      (.writeBinary gen (to-byte-array hash)))))

(def object-mapper-module
  (doto (SimpleModule. "Hash")
    (.addSerializer Hash serializer)))

(defn generate
  "Calculates a SHA256 hash for `resource`.

  The hash need to be cryptographic because otherwise it would be possible to
  introduce a resource into Blaze which has the same hash as the target
  resource, overwriting it."
  ^Hash [resource]
  (let [hasher (.newHasher (Hashing/sha256))]
    (Base/hashInto resource hasher)
    (from-byte-buffer! (bb/wrap (.asBytes (.hash hasher))))))

(defmethod print-method Hash [^Hash hash ^Writer w]
  (.write w "#blaze/hash\"")
  (.write w (.toString hash))
  (.write w "\""))

(defmethod print-dup Hash [^Hash hash ^Writer w]
  (.write w "#=(blaze.fhir.Hash/fromHex ")
  (print-dup (.toString hash) w)
  (.write w ")"))

File: modules/fhir-structure/java/blaze/fhir/Hash.java
package blaze.fhir;

import com.google.common.io.BaseEncoding;

import java.nio.ByteBuffer;

public final class Hash {

    public static final int SIZE = 32;
    public static final int PREFIX_SIZE = Integer.BYTES;
    public static final Hash DELETED = new Hash(0, 0, 0, 0);

    private final long l0;
    private final long l1;
    private final long l2;
    private final long l3;

    private Hash(long l0, long l1, long l2, long l3) {
        this.l0 = l0;
        this.l1 = l1;
        this.l2 = l2;
        this.l3 = l3;
    }

    public static Hash fromHex(String s) {
        return fromByteBuffer(ByteBuffer.wrap(BaseEncoding.base16().decode(s)));
    }

    public static Hash fromByteBuffer(ByteBuffer buffer) {
        return new Hash(buffer.getLong(), buffer.getLong(), buffer.getLong(), buffer.getLong());
    }

    public int prefix() {
        return (int) (l0 >>> 32);
    }

    public void copyTo(ByteBuffer target) {
        target.putLong(l0);
        target.putLong(l1);
        target.putLong(l2);
        target.putLong(l3);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        return o instanceof Hash that && l0 == that.l0 && l1 == that.l1 && l2 == that.l2 && l3 == that.l3;
    }

    @Override
    public int hashCode() {
        return (int) l0;
    }

    @Override
    public String toString() {
        var buffer = ByteBuffer.allocate(SIZE);
        copyTo(buffer);
        return BaseEncoding.base16().encode(buffer.array());
    }
}
